File: prime_graph.rb
module PrimeGraph

  def PrimeGraph.reload
    load 'prime_graph.rb'
  end

  class Point
    attr_accessor :next

    def inspect
      {
        has_next: !self.next.nil?
      }
    end

    def dup
      point = Point.new
      point.next = @next if @next
      point 
    end

    def to_s
      JSON.pretty_generate(self.inspect)
    end
  end

  class Cycle
    attr_reader :current_point, :initial_point, :is_complete

    def initialize(initial_point: nil, cycle: nil)
      raise "You need to pass the initial point or a cycle" unless (initial_point || cycle)
      if initial_point
        @initial_point = initial_point
        @current_point = @initial_point
      else
        @initial_point = cycle.initial_point.dup
        @current_point = @initial_point
        current = @initial_point
        while ((next_point = current.next) && next_point != cycle.initial_point)
          current = next_point.dup
          self << current
        end
      end
      @is_complete = false
    end

    def <<(point)
      raise "The cycle has already been closed" if @is_complete
      @current_point.next = point 
      @current_point = point
      self
    end

    def close_cycle
      self << @initial_point
      @is_complete = true
      self
    end

    def length
      return @length if @length
      
      current = @initial_point
      count = 1
      while ((next_point = current.next) && next_point != @initial_point)
        count += 1
        current = next_point
      end
      @length = count if @is_complete
      count
    end

    def at_starting_point?
      @current_point == @initial_point
    end

    def move(number_of_moves=1)
      (0...number_of_moves).each { @current_point = @current_point.next }
      self
    end

    def position
      position = 0
      current = @initial_point
      while (current != @current_point)
        position += 1
        current = current.next
      end
      position 
    end

    def ==(cycle)
      cycle && cycle.class == Cycle && (cycle.is_complete == @is_complete) && cycle.length == self.length
    end

    def Cycle.add(cycle1, cycle2)
      raise "Cycle is not closed" unless cycle1.is_complete && cycle2.is_complete
      raise "Incompatible cycle lengths" unless cycle1.length == cycle2.length
      
      cycle = Cycle.new cycle: cycle1
      cycle.close_cycle
      
      current = cycle1.initial_point
      while (current != cycle1.current_point)
        current = current.next
        cycle.move
      end
      current = cycle2.initial_point
      while (current != cycle2.current_point)
        current = current.next
        cycle.move
      end
  
      cycle
    end

    def inspect
      {
        length: self.length,
        position: self.position,
        is_complete: self.is_complete
      }
    end

  end

  class CycleGraph
    attr_reader :cycles

    def initialize(*cycles)
      @cycles = cycles.flatten.sort {|a,b| a.length <=> b.length } if cycles
      @cycles ||= []
    end

    def move(number_of_moves=1)
      (0...number_of_moves).each { @cycles.each(&:move) }
      self
    end

    def ==(cycle_graph)
      cycle_graph && cycle_graph.class == CycleGraph && (@cycles.map(&:size) - cycle_graph.cycles.map(&:size))
    end

    def CycleGraph.add(cycle1, cycle2)
      if cycle1.cycles.map(&:length) != cycle2.cycles.map(&:length)
        raise "Incompatible cycle lengths"
      end
      cycles = []
      (0...(cycle1.cycles.count)).each do |i|
        cycles << Cycle.add(cycle1[i], cycle2[i])
      end
      CycleGraph.new cycles
    end

    def [](i)
      @cycles[i]
    end

    def position
      @cycles.map(&:position)
    end

    def inspect
      {
        position: self.position,
        cycles: self.cycles
      }
    end

    def fully_active?
      @cycles.all?(&:at_starting_point?)
    end

    def active?
      @cycles.count > 0 && @cycles.any?(&:at_starting_point?)
    end

  end

  class PrimeGraph < CycleGraph

    attr_reader :iteration

    def initialize
      @cycles = []
      @initial_point = Point.new
      @unfinished_cycle = Cycle.new(initial_point: @initial_point)
      @unfinished_cycle << Point.new
      @iteration = 1
    end

    def run_machine(number_of_iterations)
      @iteration += number_of_iterations
      while (number_of_iterations -= 1) >= 0
        puts "Iterations left: #{number_of_iterations}"
        self.move
        unless self.active? 
          @unfinished_cycle.close_cycle
          @cycles << @unfinished_cycle
          @unfinished_cycle = Cycle.new(cycle: @unfinished_cycle)
        end
        @unfinished_cycle << Point.new
      end
      self
    end

    def inspect
      {
        iteration: self.iteration,
        primes: self.primes,
        position: self.position
      }
    end

    def primes 
      @cycles.map(&:length)
    end

  end

end

File: primeGraph.ts
class Vertex {
  public edges: object

  constructor() {
    this.edges = {}
  }

  addEdge(key: string, point) {
    this.edges[key] = point
  }
}

class Point extends Vertex {
  private _next: Point

  constructor(multipleEdges = false) {
    if (multipleEdges) super()
    this.edges = null
  }

  next(key = null) {
    return this._next || this.edges[key]
  }

  addEdge(key = null, point) {
    if (key && this.edges) super.addEdge(key, point)
    else this._next = point
    return self
  }

  dup() {
    const point = new Point()
    if (this._next) point.next = this.next
    return point
  }
}

class Cycle {
  public currentPoint: Point
  public isComplete: boolean

  private _length: number

  constructor(public initialPoint: Point = null, cycleToCopy: Cycle = null) {
    if (initialPoint && cycleToCopy) {
      console.log("Ignoring initialPoint since cycleToCopy was passed")
    }
    this.isComplete = false

    if (cycleToCopy) {
      let _initialPoint = cycleToCopy.initialPoint
      this.initialPoint = initialPoint.edges ? initialPoint : initialPoint.dup()
      this.currentPoint = this.initialPoint

      let current = this.initialPoint
      let nextPoint;
      while (nextPoint = current.next(cycleToCopy)) {
        current = nextPoint.dup()

      }

    }
  }

  addPoint(point: Point) {
    if (this.isComplete) {
      throw('The cycle has already been closed')
    }

    this.currentPoint.addEdge(this, point)
    this.currentPoint = point
    return this
  }

  closeCycle() {
    this.addPoint(this.initialPoint)
    this.isComplete = true
    return this
  }

  length() {
    if (this._length) return this._length
    let current = this.initialPoint
    let count = 1
    let nextPoint
    while ((nextPoint = current.next()) && nextPoint != this.initialPoint) {
      count++
      current = nextPoint
    }
    if (this.isComplete) this._length = count
    return count
  }

  atStartingPoint() {
    return this.currentPoint == this.initialPoint
  }

  move(numberOfMoves = 1) {
    for (let i = 0; i < numberOfMoves; i++) {
      this.currentPoint = this.currentPoint.next(self)
    }
    return this
  }

  position() {
    let position = 0
    let current = this.initialPoint
    while (current != this.currentPoint) {
      position++
      current = current.next(this)
    }
    return position
  }

  eq(cycle: Cycle) {
    return cycle && cycle instanceof Cycle && (cycle.isComplete == this.isComplete) && cycle.length() == this.length()
  }

  static add(cycle1: Cycle, cycle2: Cycle) {
    if (!(cycle1.isComplete && cycle2.isComplete)) {
      throw ('Cycle is not closed')
    }
    if (!(cycle1.length() == cycle2.length())) {
      throw ('Incompatible cycle lengths')
    }

    let cycle = new Cycle(null, cycle1)
    cycle.closeCycle()

    let current = cycle1.initialPoint
    while (current != cycle1.initialPoint) {
      current = current.next()
      cycle.move()
    }

    current = cycle2.initialPoint
    while (current != cycle2.currentPoint) {
      current = current.next()
      cycle.move()
    }

    return cycle
  }

}

class CycleGraph {
  public cycles: Cycle[] = []

  constructor(cycles: Cycle[]) {
    if (cycles && cycles instanceof Array) {
      this.cycles = cycles.sort((a, b) => a.length() - b.length())
    }
  }

  move(numberOfMoves = 1) {
    for (let i = 0; i < numberOfMoves; i++) {
      this.cycles.forEach(c => c.move())
    }
    return this
  }

  eq(cycleGraph: CycleGraph) {
    return cycleGraph && cycleGraph instanceof CycleGraph && (this.cycles.map(c => c.length()) == cycleGraph.cycles.map(c => c.length()))
  }

  static add(cycle1: CycleGraph, cycle2: CycleGraph) {
    if (cycle1.cycles.length != cycle2.cycles.length) {
      throw('Incompatible cycle lengths')
    }
    let cycles = []

    for (let i = 0; i < cycle1.cycles.length; i++) {
      cycles.push(Cycle.add(cycle1.atIndex(i), cycle2.atIndex(i)))
    }
    return new CycleGraph(cycles)
  }

  atIndex(i: number) {
    return this.cycles[i]
  }

  position() {
    return this.cycles.map(c => c.position())
  }

  fullyActive() {
    return this.cycles.every(c => c.atStartingPoint())
  }

  active() {
    return this.cycles.length > 0 && this.cycles.some(c => c.atStartingPoint())
  }
}

class PrimeGraph extends CycleGraph {
  public cycles: Cycle[]
  public initialPoint = new Point()
  public unfinishedCycle = new Cycle(this.initialPoint)
  public iteration = 1

  constructor() {
    super([])
    this.unfinishedCycle.addPoint(new Point())
  }

  runMachine(numberOfIterations) {
    this.iteration += numberOfIterations
    while ((numberOfIterations -= 1) >= 0) {
      console.log(`Iterations left: ${numberOfIterations}`)
      this.move()
      if (!this.active) {
        this.unfinishedCycle.closeCycle()
        this.cycles.push(this.unfinishedCycle)
        this.unfinishedCycle = new Cycle(null, this.unfinishedCycle)
      }
      this.unfinishedCycle.addPoint(new Point())
    }
    return this
  }

  prime() {
    return this.cycles.map(c => c.length())
  }
}